substrate-node/pallets/pallet-smart-contract/src/migrations/v12.rs

```diff
@@ -0,0 +1,200 @@
+            migrate_to_version_12::<T>()
+        } else {
+            info!(" >>> Unused Smart Contract pallet V12 migration");
+            Weight::zero()
+        }
+    }
+
+    #[cfg(feature = "try-runtime")]
+    fn post_upgrade(count: Vec<u8>) -> Result<(), sp_runtime::TryRuntimeError> {
+        validate_pallet_version::<T>(types::StorageVersion::V12)?;
+
+        let new_count = ContractPaymentState::<T>::iter().count();
+        let old_count =
+            usize::from_le_bytes(count.try_into().expect("slice with incorrect length"));
+
+        if old_count != 0 {
+            debug!(
+                "🏁  Smart Contract pallet {:?} ContractPaymentState length after migration {:?}",
+                PalletVersion::<T>::get(),
+                new_count
+            );
+            ensure!(
+                new_count == old_count,
+                DispatchError::Other(
+                    "Number of ContractPaymentState migrated does not match: {:?}"
+                )
+            );
+        }
+
+        check_contract_lock_v12::<T>()
+    }
+}
+
+#[cfg(feature = "try-runtime")]
+fn validate_pallet_version<T: Config>(
+    expected_version: types::StorageVersion,
+) -> Result<(), sp_runtime::TryRuntimeError> {
+    let current_version = PalletVersion::<T>::get();
+    debug!("current pallet version: {:?}", current_version);
+    ensure!(
+        current_version >= expected_version,
+        DispatchError::Other("Unexpected pallet version")
+    );
+    Ok(())
+}
+
+pub fn migrate_to_version_12<T: Config>() -> frame_support::weights::Weight {
+    debug!(
+        " >>> Starting contract pallet migration, pallet version: {:?}",
+        PalletVersion::<T>::get()
+    );
+
+    let mut total_reads = 0;
+    let mut total_writes = 0;
+
+    for (contract_id, contract) in Contracts::<T>::iter() {
+        log::debug!("Contract id: {:?}", contract_id);
+
+        if let Some(src_twin) = pallet_tfgrid::Twins::<T>::get(contract.twin_id) {
+            if ContractLock::<T>::contains_key(contract_id) {
+                let (r, w) = migrate_contract_lock::<T>(contract_id);
+                total_reads += r;
+                total_writes += w;
+
+                let (r, w) = remove_all_locks::<T>(&src_twin.account_id);
+                total_reads += r;
+                total_writes += w;
+            } else {
+                log::debug!("ContractLock not found for contract {:?}", contract_id);
+            }
+        } else {
+            log::debug!("Twin not found for contract {:?}", contract_id);
+        }
+    }
+
+    // Set the new storage version
+    PalletVersion::<T>::put(types::StorageVersion::V12);
+    total_writes += 1;
+
+    T::DbWeight::get().reads_writes(total_reads, total_writes)
+}
+
+fn migrate_contract_lock<T: Config>(contract_id: u64) -> (u64, u64) {
+    let mut reads = 0;
+    let mut writes = 0;
+
+    let old_contract_lock = ContractLock::<T>::take(contract_id);
```

Comment:
Do you mean we should load all contracts first and then clean the entire storage all at once?
I feel uncomfortable overloading the system by loading all the contacts into the memory during a critical process. In the example you mentioned, the expected number of items was very small.

I'm not certain if that would improve anything or if it's necessary.

But if you mean to use "get" in place of "take", then clear the storage prefix at the end, and you think this may improve the performance of the migration. I can do a test and see.

Reply:
Also, note that the example you refer to uses None for maybe_limit which is potentially unsafe since it's an unbounded operation.

Reply:
The "clear_storage_prefix" function involves iterating over the items in the storage map. If we first read the values and then remove them which involves both read and write operations, it would lead to a greater number of total reads and worse performance.

However, I refactored the migration in a way that improved the overall performance of the migration.

```
[2024-08-27T00:14:54Z INFO  pallet_smart_contract::migrations::v12]  >>> Unused Smart Contract pallet V12 migration
[2024-08-27T00:14:54Z INFO  try-runtime::cli] PoV size (zstd-compressed compact proof): 218.9 KB. For parachains, it's your responsibility to verify that a PoV of this size fits within any relaychain constraints.
[2024-08-27T00:14:54Z INFO  try-runtime::cli] Consumed ref_time: 0.887775s (44.39% of max 2s)
[2024-08-27T00:14:54Z INFO  try-runtime::cli] ✅ No weight safety issues detected. Please note this does not guarantee a successful runtime upgrade. Always test your runtime upgrade with recent state, and ensure that the weight usage of your migrations will not drastically differ between testing and actual on-chain execution.
```

Also, it handles the extra case where a user may have no contracts at the time of the migration, and also has leftover locks in his account.
So instead of looping over the conatrcts and getting the accounts, we loop over the balance locks, get the accounts and clear the locks.

Reply:
> So instead of looping over the conatrcts and getting the accounts, we loop over the balance locks, get the accounts and clear the locks.

Great! I ll check this